src/NServiceBus.Persistence.DynamoDB/Outbox/WriteRequestBatchExecutionExtensions.cs

```diff
@@ -0,0 +1,178 @@
+                    throw;
+                }
+            }
+
+
+            if (!succeeded)
+            {
+                logger.Warn($"(5 / {maximumNumberOfRetries}) All retry attempts for batch '{batchNumber}/{totalBatches}' with entries '{CreateBatchLogMessage(batch, configuration)}' to table '{configuration.Table.TableName}' exhausted.");
+            }
+        }
+
+        static async Task<BatchWriteItemResponse> WriteBatch(this IAmazonDynamoDB dynamoDbClient,
+            List<WriteRequest> batch, int batchNumber, int totalBatches,
+            OutboxPersistenceConfiguration configuration, ILog logger, CancellationToken cancellationToken)
+        {
+            string? logBatchEntries = null;
+            if (logger.IsDebugEnabled)
+            {
+                logBatchEntries = CreateBatchLogMessage(batch, configuration);
+                logger.Debug(
+                    $"Writing batch '{batchNumber}/{totalBatches}' with entries '{logBatchEntries}' to table '{configuration.Table.TableName}'");
+            }
+
+            var batchWriteItemRequest = new BatchWriteItemRequest
+            {
+                RequestItems =
+                    new Dictionary<string, List<WriteRequest>> { { configuration.Table.TableName, batch } },
+            };
+
+            var result = await dynamoDbClient.BatchWriteItemAsync(batchWriteItemRequest, cancellationToken)
+                .ConfigureAwait(false);
+
+            if (logger.IsDebugEnabled)
+            {
+                logger.Debug(
+                    $"Wrote batch '{batchNumber}/{totalBatches}' with entries '{logBatchEntries}' to table '{configuration.Table.TableName}'");
+            }
+
+            return result;
+        }
+
+        static string CreateBatchLogMessage(IReadOnlyCollection<WriteRequest> batch, OutboxPersistenceConfiguration configuration)
+        {
+            var stringBuilder = new StringBuilder();
+
+            foreach (var writeRequest in batch)
+            {
+                if (writeRequest.DeleteRequest is { } deleteRequest)
+                {
+                    stringBuilder.Append($"DELETE #PK {deleteRequest.Key[configuration.Table.PartitionKeyName].S} / #SK {deleteRequest.Key[configuration.Table.SortKeyName].S}, ");
+                }
+
+                if (writeRequest.PutRequest is { } putRequest)
+                {
+                    stringBuilder.Append($"PUT #PK {putRequest.Item[configuration.Table.PartitionKeyName].S} / #SK {putRequest.Item[configuration.Table.SortKeyName].S}, ");
+                }
+            }
+
+            return stringBuilder.Length > 2 ? stringBuilder.ToString(0, stringBuilder.Length - 2) : stringBuilder.ToString();
+        }
+
+        static TimeSpan CalculateDelay(int attempt, TimeSpan? retryDelay)
+        {
+            var delay = Convert.ToInt32(retryDelay.GetValueOrDefault(TimeSpan.FromMilliseconds(250)).TotalMilliseconds);
+            return TimeSpan.FromMilliseconds(attempt * delay);
```

Comment:
do we really want to increase the delay time?

Reply:
Yes. It is explicitly mentioned to be required by the documentation. I will add a comment

> If DynamoDB returns any unprocessed items, you should retry the batch operation on those items. However, we strongly recommend that you use an exponential backoff algorithm. If you retry the batch operation immediately, the underlying read or write requests can still fail due to throttling on the individual tables. If you delay the batch operation using exponential backoff, the individual requests in the batch are much more likely to succeed.

https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Programming.Errors.html#Programming.Errors.BatchOperations

Although I didn't use an exponential backoff here.